### PR TITLE
overlord/ifacestate: track connections in the state

### PR DESCRIPTION
**Pull request by zyga**
This branch ensures that connections are tracked in the state

Signed-off-by: Zygmunt Krynicki <zygmunt.krynicki@canonical.com>

### CODE REVIEW

File: overlord/ifacestate/ifacemgr.go [hunk 1]
```diff
@@ -277,6 +277,24 @@ func getPlugAndSlotRefs(task *state.Task) (*interfaces.PlugRef, *interfaces.Slot
 	return &plugRef, &slotRef, nil
 }
 
+func getConns(task *state.Task) (map[string]connState, error) {
+	// Get information about connections from the state
+	var conns map[string]connState
+	err := task.State().Get("conns", &conns)
+	if err != nil && err != state.ErrNoState {
+		task.Errorf("cannot obtain data about existing connections, %s", err)
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: s/,/:/, as usual elsewhere

Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Ah, and we don't need to call this here. Just return the error with fmt.Errorf.

File: overlord/ifacestate/ifacemgr.go [hunk 2]
```diff
@@ -277,6 +277,24 @@ func getPlugAndSlotRefs(task *state.Task) (*interfaces.PlugRef, *interfaces.Slot
 	return &plugRef, &slotRef, nil
 }
 
+func getConns(task *state.Task) (map[string]connState, error) {
+	// Get information about connections from the state
+	var conns map[string]connState
+	err := task.State().Get("conns", &conns)
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Feels strange to have task here. This function should take a *State instead.

File: overlord/ifacestate/ifacemgr.go [hunk 3]
```diff
@@ -277,6 +277,24 @@ func getPlugAndSlotRefs(task *state.Task) (*interfaces.PlugRef, *interfaces.Slot
...
+		task.Errorf("cannot obtain data about existing connections, %s", err)
+		return nil, err
+	}
+	if conns == nil {
+		conns = make(map[string]connState)
+	}
+	return conns, nil
+}
+
+func setConns(task *state.Task, conns map[string]connState) {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Same, let's please have the *State here rather than a task.

File: overlord/ifacestate/ifacemgr.go [hunk 4]
```diff
@@ -277,6 +277,24 @@ func getPlugAndSlotRefs(task *state.Task) (*interfaces.PlugRef, *interfaces.Slot
...
+	task.State().Set("conns", conns)
+}
+
 func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
 	task.State().Lock()
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: st := task.State()

So we can use it.

File: overlord/ifacestate/ifacemgr.go [hunk 5]
```diff
@@ -285,7 +303,21 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
...
 		return err
 	}
-	return m.repo.Connect(plugRef.Snap, plugRef.Name, slotRef.Snap, slotRef.Name)
+
+	conns, err := getConns(task)
+	if err != nil {
+		return err
+	}
+
+	if err := m.repo.Connect(plugRef.Snap, plugRef.Name, slotRef.Snap, slotRef.Name); err != nil {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Let's please break this line down so the code is more readable (err != nil on its own).

File: overlord/ifacestate/ifacemgr.go [hunk 6]
```diff
@@ -285,7 +303,21 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
...
+		return err
+	}
+
+	plug := m.repo.Plug(plugRef.Snap, plugRef.Name)
+	connID := fmt.Sprintf("%s:%s %s:%s", plugRef.Snap, plugRef.Name, slotRef.Snap, slotRef.Name)
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Let's please have a function instead:

`func connID(plug, slot) string`

and use it below directly:

`conns[connID(plugRef, slotRef)] = connState{Interface: plug.Interface}`

`Auto` can go as it's the default.

File: overlord/ifacestate/ifacemgr.go (hunk 2 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: The reason it takes a task is that task can log things in a way that will be hopefully useful. Should I just not log anything at all or use a different logging mechanism?

File: overlord/ifacestate/ifacemgr.go (hunk 1 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Ok, that answers the question above too

File: overlord/ifacestate/ifacemgr.go [hunk 7]
```diff
@@ -285,7 +303,21 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
...
+	conns[connID] = connState{Interface: plug.Interface, Auto: false}
+	setConns(task, conns)
+	return nil
 }
 
 func (m *InterfaceManager) doDisconnect(task *state.Task, _ *tomb.Tomb) error {
```
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: Similar changes to doConnect.

**Comment by niemeyer** [comment]
A few simple changes, and needs tests.

File: overlord/ifacestate/ifacemgr.go (hunk 2 above)
Comment (niemeyer) [inline overlord/ifacestate/ifacemgr.go]: The exact Errorf line is done by the TaskRunner.

File: overlord/ifacestate/ifacemgr.go (hunk 6 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go (hunk 5 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go (hunk 7 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go (hunk 4 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

File: overlord/ifacestate/ifacemgr.go (hunk 3 above)
Comment (zyga) [inline overlord/ifacestate/ifacemgr.go]: Done

**Pull request merged**